storage/engine: add Pebble support to ExportToSst

hueypark opened the pull request:
Most of these commits are from [engine: Add ExportToSst method with all logic moved to C++](https://github.com/cockroachdb/cockroach/commit/75f1314b9ad4326069784838c581e1cb7a5ede24).

Fixes #41739

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42134)
<!-- Reviewable:end -->

itsbilal reviewed (commented):
Great work so far! This looks pretty solid overall - just some relatively minor comments. Thanks for doing this!

> Reviewed 18 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-:-LssJMRKESWm5zEFPvln:b-mho6s4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark)

___
*[pkg/ccl/storageccl/export_test.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssJae8DhEBFreuI1WM:-LssJae8DhEBFreuI1WN:b-flryz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/320c1aad52af8a31afd1d5ef86a95e91aac7a446/pkg/ccl/storageccl/export_test.go#L230)):*
> ```Go
> 		EndTime:                             endTime,
> 		UpperBound:                          endKey,
> 		EnableTimeBoundIteratorOptimization: enableTimeBoundIteratorOptimization,
> ```

I see that `enableTimeBoundIteratorOptimization` is a parameter of this method, but `IterOptions.{Min, Max}TimestampHint` isn't being set accordingly if it's true. Maybe add that check right before `NewMVCCIncrementalIterator` is called?
___
*[pkg/storage/bulk/sst_batcher.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssM2A4CTbCqMXCALRx:-LssM2A4CTbCqMXCALRy:b-4mqshb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/bulk/sst_batcher.go#L341)):*
> ```Go
> }
> 
> // SSTSender is the sst sender interface to an engine's data.
> ```

Nit: Probably not the best idea to describe an interface with its name alone, maybe "SSTSender is an interface to send SST data to an engine" ?
___
*[pkg/storage/engine/pebble_batch.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssQYZR99sDIGWefhpo:-LssQYZR99sDIGWefhpp:b-wzaou7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/pebble_batch.go#L95)):*
> ```Go
> 	start, end MVCCKey, exportAllRevisions bool, io IterOptions,
> ) ([]byte, roachpb.BulkOpSummary, error) {
> 	return pebbleExportToSst(p, start, end, exportAllRevisions, io)
> ```

While this works, it's also worthwhile to just panic here for consistency with rocksdb. See comment below.
___
*[pkg/storage/engine/rocksdb.go, line 1711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssQAiZEqirv12N90TG:-LssQAiZEqirv12N90TH:b-x2y0uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/rocksdb.go#L1711)):*
> ```Go
> 	start, end MVCCKey, exportAllRevisions bool, io IterOptions,
> ) ([]byte, roachpb.BulkOpSummary, error) {
> 	return r.parent.ExportToSst(start, end, exportAllRevisions, io)
> ```

Looks like this is just a passthrough onto the parent db instance's `ExportToSst`, potentially skipping over mutations in this batch. The previous code didn't support `ExportToSst` on a rocksdb batch anyway, so it's probably worthwhile to just do a `panic("unimplemented")` here in case we ever unintentionally exercise this functionality.
___
*[pkg/storage/engine/sst_writer.go, line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssQtMlD3F8KczbfIgo:-LssQtMlD3F8KczbfIgp:b-44ntnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/sst_writer.go#L151)):*
> ```Go
> // MakeRocksSST creates RocksDB SST.
> func MakeRocksSST(kvs []MVCCKeyValue) ([]byte, error) {
> 	w, err := MakeRocksDBSstFileWriter()
> ```

I see this was copied over from testing code two different files, one is now in `bulk` and the other is in `engine`.. But I'm not sure how good of an idea it is to make it an exported non-test function - it's only called in testing code, plus we want to remove usage of `MakeRocksDBSstFileWriter`. I'd say move these two functions back to the test files (even if it's duplicated in both of them) and make them unexported?


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42134)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark)

___
*[pkg/storage/engine/engine.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LswJ2em25NVAOfG05iM:-LswJ2em25NVAOfG05iN:b-2rshh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/engine.go#L193)):*
> ```Go
> 	// SSTable containing the exported keys, the size of exported data, or an error.
> 	ExportToSst(
> 		start, end MVCCKey, exportAllRevisions bool, io IterOptions,
> ```

The semantics of the range exported are a bit subtle. The fact that `start.Key` is inclusive while `start.Timestamp` is exclusive could be confusing. I think it might be better to have separate `startKey,endKey roachpb.Key` and `startTS, endTS hlc.Timestamp` options.  Rather than make this change in this PR, can you add a TODO?


<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
Thank you for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-:-Lsvk6p799Fu2logKgjO:b-rrkdbt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @petermattis)

___
*[pkg/ccl/storageccl/export_test.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssJae8DhEBFreuI1WM:-Lsvll0-4HtBYD83UBKD:bhmhk5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/320c1aad52af8a31afd1d5ef86a95e91aac7a446/pkg/ccl/storageccl/export_test.go#L230)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I see that `enableTimeBoundIteratorOptimization` is a parameter of this method, but `IterOptions.{Min, Max}TimestampHint` isn't being set accordingly if it's true. Maybe add that check right before `NewMVCCIncrementalIterator` is called?
</blockquote></details>

Nice catch! Done.
___
*[pkg/storage/bulk/sst_batcher.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssM2A4CTbCqMXCALRx:-LsvlzEj3BTwEijKIax_:b-i21bbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/bulk/sst_batcher.go#L341)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Nit: Probably not the best idea to describe an interface with its name alone, maybe "SSTSender is an interface to send SST data to an engine" ?
</blockquote></details>

Thanks. Done.
___
*[pkg/storage/engine/engine.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LswJ2em25NVAOfG05iM:-Lt-SZ29BC1NPt6W6mP9:bffzkuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/engine.go#L193)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The semantics of the range exported are a bit subtle. The fact that `start.Key` is inclusive while `start.Timestamp` is exclusive could be confusing. I think it might be better to have separate `startKey,endKey roachpb.Key` and `startTS, endTS hlc.Timestamp` options.  Rather than make this change in this PR, can you add a TODO?
</blockquote></details>

I added TODO.
___
*[pkg/storage/engine/pebble_batch.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssQYZR99sDIGWefhpo:-LsvmTI_DwYjFwfRvozT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/pebble_batch.go#L95)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

While this works, it's also worthwhile to just panic here for consistency with rocksdb. See comment below.
</blockquote></details>

Done.
___
*[pkg/storage/engine/rocksdb.go, line 1711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssQAiZEqirv12N90TG:-LsvmNkBETsQzxydvgOE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/rocksdb.go#L1711)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Looks like this is just a passthrough onto the parent db instance's `ExportToSst`, potentially skipping over mutations in this batch. The previous code didn't support `ExportToSst` on a rocksdb batch anyway, so it's probably worthwhile to just do a `panic("unimplemented")` here in case we ever unintentionally exercise this functionality.
</blockquote></details>

Done.
___
*[pkg/storage/engine/sst_writer.go, line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssQtMlD3F8KczbfIgo:-Lt-OjeX3uUKv6qfkByw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/engine/sst_writer.go#L151)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I see this was copied over from testing code two different files, one is now in `bulk` and the other is in `engine`.. But I'm not sure how good of an idea it is to make it an exported non-test function - it's only called in testing code, plus we want to remove usage of `MakeRocksDBSstFileWriter`. I'd say move these two functions back to the test files (even if it's duplicated in both of them) and make them unexported?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> save for two really minor suggestions. Thanks so much!

> Reviewed 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-:-Lt1Mmw46M_Jqx8Ox0NN:b-nwxe0p)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @hueypark, @itsbilal, and @petermattis)

___
*[pkg/ccl/storageccl/export_test.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-Lt1MKC34KCgWeJLixsK:-Lt1MKC443GAm8tfBsBY:baydbmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/facb92793424c991061cfd0b10e571196917c46f/pkg/ccl/storageccl/export_test.go#L231)):*
> ```Go
> 	if enableTimeBoundIteratorOptimization {
> 		io.MaxTimestampHint = endTime
> 		io.MinTimestampHint = startTime
> ```

While this works too, `startTime.Next()` is more correct here- and matches what the old code did.
___
*[pkg/storage/bulk/sst_batcher.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-LssM2A4CTbCqMXCALRx:-Lt1MYxs43wrBJAOhas1:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8a5193cd1cc3123d5df495479c0dc23855a0ede/pkg/storage/bulk/sst_batcher.go#L341)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

Thanks. Done.
</blockquote></details>

:+1: 
___
*[pkg/storage/engine/engine.go, line 192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-Lt1K58O6U9tGlP_UuQS:-Lt1K58O6U9tGlP_UuQT:b-qwhl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/facb92793424c991061cfd0b10e571196917c46f/pkg/storage/engine/engine.go#L192)):*
> ```Go
> 	// requested or if the start.Timestamp is non-zero. Returns the bytes of an
> 	// SSTable containing the exported keys, the size of exported data, or an error.
> 	// TODO(hueypark): Separate MVCCKey into roachpb.Key and hlc.Timestamp.
> ```

Nit: add a blank line before the TODO to make it more obvious.


<!-- Sent from Reviewable.io -->

hueypark committed:
storage/engine: add Pebble support to ExportToSst

Most of these commits are from [engine: Add ExportToSst method with all logic moved to C++](https://github.com/cockroachdb/cockroach/commit/75f1314b9ad4326069784838c581e1cb7a5ede24).

Fixes #41739

Release note: None

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42134)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @petermattis)

___
*[pkg/ccl/storageccl/export_test.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-Lt1MKC34KCgWeJLixsK:-Lt1iFbm6MVrvDn8v7Fv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/facb92793424c991061cfd0b10e571196917c46f/pkg/ccl/storageccl/export_test.go#L231)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

While this works too, `startTime.Next()` is more correct here- and matches what the old code did.
</blockquote></details>

Done.
___
*[pkg/storage/engine/engine.go, line 192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-Lt1K58O6U9tGlP_UuQS:-Lt1ga3YDzEsC-uD8Pzv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/facb92793424c991061cfd0b10e571196917c46f/pkg/storage/engine/engine.go#L192)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Nit: add a blank line before the TODO to make it more obvious.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

I'll go ahead and have this merged now, thanks!

bors r+

> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42134#-:-Lt2IxGbEb3Hk_ZYDw_N:buz5g4g)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @petermattis)



<!-- Sent from Reviewable.io -->

petermattis commented:
Unrelated CI flake. Retrying.

bors r+
